#32 page with data from server

mcalthrop opened the pull request:
Implements #32.

Note: the page changed is _Edit Profile_; at this point, the data displayed on that page is read-only.

Commit:
Rename class of ProfileEditPage

Was previously incorrect: NotFoundView

Commit:
Remove unused .gitkeep file

Commit:
Update edit profile page

- create component: `ProfileEdit`
- requires `user` property
- displays relevant values from `user`
- rewire `ProfileEditPage` to use `ProfileEdit`

Commit:
Add API call to get user profile

Commit:
Various improvements, inc API call

- add flow annotations
- refactor auth process
- implement async action creators

Commit:
Fix build task

The `npm run build` script did not copy static files.
Also added a clean of `dist` dir.

Commit:
Update flow annotations to use "Maybe" types

See http://flowtype.org/docs/nullable-types.html

Commit:
Remove dotenv from package.json

mcalthrop commented:
**NOTE:** not yet ready for merging; wanted to trigger a Travis build to check it.

Commit:
First attempt at wiring in an *Update* button

On the *Edit Profile* page

Commit:
Fix *Update* button click handling

Commit:
Rename `LanguageSelectionDropDown`

To `LanguageSelectionDropdown`

Commit:
Add Action dispatch when profile updated

Commit:
Separate user update into different actions

- UPDATE_PROFILE_REQUEST
- UPDATE_PROFILE_SUCCESS
- UPDATE_PROFILE_FAILURE

Commit:
Document how to mock service call failure

Commit:
Interim commit: spinner in modal overlay

Still things to sort out, as described in TODOs

Commit:
Use Symbol for CHANGE_LANGUAGE action

Commit:
Fix flow error in language

Commit:
Wire in Spinner to Layouts

Add to AdminPageLayout and HeroPageLayout.

Commit:
Rename IndeterminateSpinner to Spinner

Commit:
Add short delay when logging in

Allows display of spinner with message

Commit:
User Symbol for LOGOUT_REQUEST

Commit:
Implement LOGOUT_SUCCESS action

Incorporates Spinner to show message (short timeout)

Commit:
Add i18n for Spinner

Commit:
Add i18n for Account and Log out

vincemtnz commented on src/redux/modules/spinner/spinner.js:
```diff
@@ -0,0 +1,97 @@
+  LOGIN_FAILURE,
+  LOGOUT_REQUEST,
+  LOGOUT_SUCCESS,
+} from 'redux/modules/auth/auth-actions';
```
Isn't it easier to `dispatch(showSpinner(message))`?
With this approach you handle all request, success, and failures, for _every_ async action here. Leaves two places to refactor when you refactor redux modules.

mcalthrop replied:
I wasn't sure how best to implement this, but your suggestion sounds good (using `messageId` as the param to the new `showSpinner()` method).

Commit:
Delete unused layouts

- ProfileLayout
- ProfileLayoutHeader

vincemtnz commented on src/redux/modules/spinner/spinner.js:
```diff
@@ -0,0 +1,97 @@
+      newState = Object.assign({}, action.state, DEFAULT_SPINNER_STATE);
+      break;
+    default:
+      newState = state;
```
This leaves you with a [Big Ball of Mud reducer](https://twitter.com/mattpodwysocki/status/689830899781763072) :confused:

mcalthrop replied:
Yeah, it could get quite big when other notifications for other service calls are required.

Solution?

vincemtnz replied:
Since spinners _always_ happen during async actions, I see the action creators as a candidate. In reality, when we do a `LOGIN_REQUEST` action, we only do it to update a loading state of some sort (although a local one, perhaps just used in a login widget). It makes sense to dispatch a global  loading state action at the same time. In `loginRequest`, at the same time you dispatch the `LOGIN_REQUEST` action, you dispatch another action: https://github.com/mcalthrop/opencredo-react-boilerplate-1/blob/feature/32-page-with-data-from-server/src/redux/modules/auth/auth-actions.js#L78-L97

```
    import { showSpinner, hideSpinner } from 'redux/models/spinner'; // <-- import our action creators

    ...

    dispatch(showSpinner(message)); // <-- show your spinner here
    dispatch({
      type: LOGIN_REQUEST,
      state: {
        isLoading: true,
        isAuthenticated: false,
      },
    });

    ...
    // do your api call and handle the promise
    .then(response => {
      dispatch(loginSuccess());
      dispatch(hideSpinner()); // <-- hide your spinner here
    }, failure => {
      dispatch(loginFailure()); // <-- hide your spinner here
      dispatch(hideSpinner());
    });
```

That way, actionCreators are _declarative_ about what's going to happen in the application.

mcalthrop replied:
Excellent. Thanks for that. I had a go before I read this, and have implemented pretty much the same solution. :smile_cat:

Commit:
Refactor to remove big ball of mud

The spinner reducer now uses its own actions:
- SHOW_SPINNER
- HIDE_SPINNER

These are dispatched at the relevant points in the code.

Commit:
Fix minor issue with Spinner message

Commit:
Refactor spinner reducer

mcalthrop commented:
@eyko: ready to check now!

vincemtnz commented on src/redux/modules/spinner/spinner.js:
```diff
@@ -0,0 +1,63 @@
+  };
+}
+
+export function spinnerReducer(state : SpinnerState = DEFAULT_SPINNER_STATE, action: any): any {
```
Reducers always return state, so `SpinnerState` in this case since you already have it.

mcalthrop replied:
Ok, done.

Commit:
Return type SpinnerState

mcalthrop commented on src/components/AdminPageLayout.js:
```diff
@@ -10,12 +13,17 @@ const AdminPageLayout = (props) => {
 };
 
-export default AdminPageLayout;
+export default connect(mapStateToProps)(AdminPageLayout);
```
With these modifications, this is no longer a **stateless component**, because it now depends on the redux store – so it needs to be moved to the `containers` directory.

mcalthrop commented on src/components/HeroPageLayout.js:
```diff
@@ -11,12 +13,17 @@ const HeroPageLayout = (props) => {
 };
 
-export default HeroPageLayout;
+export default connect(mapStateToProps)(HeroPageLayout);
```
Same as for `AdminPageLayout`: need to move to `containers` directory.
